### PR TITLE
feat: added ability to make a singleShare to instagram stories on And…

### LINKED ISSUE #712: Instagram Stories for Android
Does the latest version 3.0.0 support Instagram Stories feature for Android?
And for ios, how do i share a video in instagram stories and also the attributionURL?

{
method: Share.InstagramStories.SHARE_BACKGROUND_IMAGE,
backgroundBottomColor: '#fefefe',
backgroundTopColor: '#906df4',
attributionURL : '',
backgroudImage : '',
social: Share.Social.INSTAGRAM_STORIES
}

The above code works fine for sharing image on instagram stories for ios.

### PR DESCRIPTION
**Pull request by vovka-s**
#Overview
This PR enables sharing of the images to instagram stories on Android
fixes #712 

# Test Plan
- try to share smth to instagram stories

### CODE REVIEW

File: android/src/main/java/cl/json/social/InstagramStoriesShare.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.facebook.react.bridge.ReadableMap;
+
+/**
+ * Created by Vladimir Stalmakov on 01-06-20.
+ */
+public class InstagramStoriesShare extends SingleShareIntent {
+
+    private static final String PACKAGE = "com.instagram.android";
+    private static final String PLAY_STORE_LINK = "market://details?id=com.instagram.android";
```
Comment (MateusAndrade) [inline android/src/main/java/cl/json/social/InstagramStoriesShare.java:18]: Why we don't use this url "https://play.google.com/store/apps/details?id=com.instagram.android" instead?

File: android/src/main/java/cl/json/social/ShareIntent.java [hunk 2]
```diff
@@ -124,6 +127,68 @@ public void open(ReadableMap options) throws ActivityNotFoundException {
...
+                        if (ShareIntent.hasValidKey("backgroundImage", options)) {
+                            this.backgroundAsset = new ShareFile(options.getString("backgroundImage"), "image/jpeg", "background", this.reactContext);
+                            this.getIntent().setDataAndType(backgroundAsset.getURI(), backgroundAsset.getType());
+                            this.getIntent().addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
+
+                            if (ShareIntent.hasValidKey("attributionURL", options)) {
+                                this.getIntent().putExtra("content_url", options.getString("attributionURL"));
+                            }
+                        } else {
+                            throw new java.lang.IllegalArgumentException("backgroundImage is empty");
```
Comment (MateusAndrade) [inline android/src/main/java/cl/json/social/ShareIntent.java:144]: We could send an error message instead. By doing this, the app will crash, right?

File: android/src/main/java/cl/json/social/SingleShareIntent.java [hunk 3]
```diff
@@ -58,13 +58,20 @@ public void open(ReadableMap options) throws ActivityNotFoundException {
...
 
-    protected void openIntentChooser() throws ActivityNotFoundException {
+    protected void openIntentChooser(ReadableMap options) throws ActivityNotFoundException {
         if (this.options.hasKey("forceDialog") && this.options.getBoolean("forceDialog")) {
             Activity activity = this.reactContext.getCurrentActivity();
             if (activity == null) {
                 TargetChosenReceiver.sendCallback(false, "Something went wrong");
                 return;
             }
+            String socialType = options.getString("social");
```
Comment (mikehardy) [inline android/src/main/java/cl/json/social/SingleShareIntent.java:68]: seems like you could avoid touching all the other share objects if you checked for null options first here, then kept the empty-argument openIntentChooser() and just chained it to this one passing null e.g.

```java
protected void openIntentChooser() throws ActivityNotFoundException {
    this.openIntentChooser(null);
}
```

That's a minor point though

File: android/src/main/java/cl/json/social/SingleShareIntent.java [hunk 4]
```diff
@@ -58,13 +58,20 @@ public void open(ReadableMap options) throws ActivityNotFoundException {
...
+            if (socialType.equals("instagramstories")) {
+                String method = options.getString("method");
+                if (method.equals("shareStickerImage") || method.equals("shareBackgroundAndStickerImage")) {
+                    activity.grantUriPermission("com.instagram.android", this.stickerAsset.getURI(), Intent.FLAG_GRANT_READ_URI_PERMISSION);
```
Comment (mikehardy) [inline android/src/main/java/cl/json/social/SingleShareIntent.java:72]: nice!

File: index.js [hunk 5]
```diff
@@ -211,7 +211,7 @@ class RNShare {
     TWITTER: NativeModules.RNShare.TWITTER || 'twitter',
     WHATSAPP: NativeModules.RNShare.WHATSAPP || 'whatsapp',
     INSTAGRAM: NativeModules.RNShare.INSTAGRAM || 'instagram',
-    INSTAGRAM_STORIES: NativeModules.RNShare.INSTAGRAM_STORIES || 'instagram-stories',
+    INSTAGRAM_STORIES: NativeModules.RNShare.INSTAGRAM_STORIES || 'instagramstories',
```
Comment (mikehardy) [inline index.js:214]: is this a breaking change?

File: android/src/main/java/cl/json/social/WhatsAppShare.java [hunk 6]
```diff
@@ -20,7 +20,7 @@ public WhatsAppShare(ReactApplicationContext reactContext) {
     public void open(ReadableMap options) throws ActivityNotFoundException {
         super.open(options);
         //  extra params here
-        this.openIntentChooser();
+        this.openIntentChooser(options);
```
Comment (MateusAndrade) [inline android/src/main/java/cl/json/social/WhatsAppShare.java:23]: Also, why we need this change on every call to the intentChooser? 🤔

File: android/src/main/java/cl/json/social/InstagramStoriesShare.java (hunk 1 above)
Comment (vovka-s) [reply android/src/main/java/cl/json/social/InstagramStoriesShare.java:18]: I used a link like from InstagramShare.java. Fixed

File: android/src/main/java/cl/json/social/ShareIntent.java (hunk 2 above)
Comment (vovka-s) [reply android/src/main/java/cl/json/social/ShareIntent.java:144]: No. This exception will be cached in RNShareModule.java.
Error message will be send from there

File: android/src/main/java/cl/json/social/InstagramStoriesShare.java (hunk 1 above)
Comment (MateusAndrade) [reply android/src/main/java/cl/json/social/InstagramStoriesShare.java:18]: I get it, maybe we can change this later. I think redirecting to Google Play is most common now.

**Comment by MateusAndrade** [comment]
This looks pretty good @vovka-s, I need to test that on my device but I think we can follow with that after these tests. Also, it would be good to change the docs regarding that we support this on Android too. 😄

File: android/src/main/java/cl/json/social/SingleShareIntent.java (hunk 3 above)
Comment (vovka-s) [reply android/src/main/java/cl/json/social/SingleShareIntent.java:68]: Good idea! I'll add the check. Thanks!

File: index.js (hunk 5 above)
Comment (vovka-s) [reply index.js:214]: No, It's only renaming of internal variable to use it as a key in enum SHARES

**Comment by vovka-s** [comment]
Thanks))
I'll change the docs in one hour

File: android/src/main/java/cl/json/social/SingleShareIntent.java [hunk 7]
```diff
@@ -58,13 +58,22 @@ public void open(ReadableMap options) throws ActivityNotFoundException {
         super.open(options);
     }
 
-    protected void openIntentChooser() throws ActivityNotFoundException {
+    protected void openIntentChooser(ReadableMap options) throws ActivityNotFoundException {
```
Comment (mikehardy) [inline android/src/main/java/cl/json/social/SingleShareIntent.java:61]: I think (I'm not sure, I haven't compiled it and I'm not sure there is android compile test in CI?) that you *must* maintain the previous API in existence as well, just chaining into this new one with null sent in as the option

That way the internal API openIntentChooser() is still backwards compatible for all the existing code (no need to update all of them) but you have options available if they were passed in and non null

Comment (vovka-s) [reply android/src/main/java/cl/json/social/SingleShareIntent.java:61]: Sorry but I don't understand how to implement this in that case. The new code is on the middle of the function, so I can't simply chain two calls

Comment (mikehardy) [reply android/src/main/java/cl/json/social/SingleShareIntent.java:61]: Does the code compile currently? I'm guessing if you changed openIntentChooser() to take an argument now as openIntentChooser(ReadableMap options) that all previous code that does not send it an argument will now fail to compile.

So you need to maintain the old API, with an implementation that just calls the new one with a null supplied as the argument.

I believe I specified an actual code suggestion in my first comment about it

Comment (mikehardy) [reply android/src/main/java/cl/json/social/SingleShareIntent.java:61]: https://github.com/react-native-community/react-native-share/pull/790#discussion_r433907751

**Comment by MateusAndrade** [comment]
Hey @vonovak, do you need any help with these conflicts?

**Comment by vovka-s** [comment]
@MateusAndrade
Sorry, what conflicts are you about?
- compilation passed
- all tests passed
- test app build passed

**Comment by friederbluemle** [comment]
@vovka-s - He's probably referring to the conflict in `ShareIntent.java` due to the changes merged in #750. When you rebase your branch you'll notice the conflict:

```
Could not apply 597af19... feat: added ability to make a singleShare to instagram stories on Android
```

In this case it's easy to resolve without the need to manually edit the file. 

```
git checkout --theirs android/src/main/java/cl/json/social/ShareIntent.java
```

Then add the file and rebase --continue.

@MateusAndrade
That's a funny autocomplete-typo, I guess you meant `@vovka-s` when you said `@vonovak` 😄

**Comment by MateusAndrade** [comment]
@vovka-s do you need help with this? Recently I changed a lot of things in the `example` app and on the documentation. You can read more about that [here](https://react-native-community.github.io/react-native-share/docs/contributing). Let me know if you have any doubt. 😄

**Comment by vovka-s** [comment]
@MateusAndrade
Sorry, I forgot about this issue.
Merge done, thanks

**Comment by MateusAndrade** [comment]
> @MateusAndrade
> Sorry, I forgot about this issue.
> Merge done, thanks

No problem, I saw that we will need to update the docs due to this change. Can you check this out?

**Comment by vovka-s** [comment]
Sure
New docs became more beautiful :+1:

**Pull request merged**